integrationtests/tests/tests.rs

```diff
@@ -188,10 +191,133 @@ async fn lightning_gateway_pays_invoice() {
+        .server
+        .await_preimage_decryption(txid)
+        .await
+        .unwrap();
+
+    // Check that the preimage matches user pubkey & lightning invoice preimage
+    let pubkey = PublicKey::from_keypair(&secp(), &keypair);
+    assert_eq!(pubkey, preimage.0);
+    assert_eq!(&sha256(&pubkey.serialize()), invoice.payment_hash());
+
+    // User claims their ecash
+    user.client
+        .claim_incoming_contract(contract_id, keypair, rng())
+        .await
+        .unwrap();
+    fed.run_consensus_epochs(4).await;
+
+    // User fetches their coins
+    user.client.fetch_all_coins().await.unwrap();
+    fed.run_consensus_epochs(2).await;
+
+    // Ecash tokens have been transferred from gateway to user
+    assert_eq!(gateway.user.client.coins().amount(), sats(0));
+    assert_eq!(user.total_coins(), amount);
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn receive_lightning_payment_invalid_preimage() {
+    let amount = sats(2000);
+    let (fed, user, _, gateway, _) = fixtures(2, 0, &[sats(10), sats(1000)]).await;
+    fed.mint_coins_for_user(&gateway.user, amount).await;
+    assert_eq!(user.total_coins(), sats(0));
+    assert_eq!(gateway.user.total_coins(), amount);
+
+    // Manually contstruct offer w/ invalid preimage. Can't use library code because it uses correct preimage ;)
+    let (fake_keypair, fake_pubkey) = secp().generate_schnorrsig_keypair(&mut rng());
+    let (real_keypair, real_pubkey) = secp().generate_schnorrsig_keypair(&mut rng());
+    let payment_hash = sha256(&real_pubkey.serialize());
+    let offer = IncomingContractOffer {
+        amount,
+        hash: payment_hash,
+        encrypted_preimage: EncryptedPreimage::new(
+            fake_pubkey.serialize(),
+            &user.config.ln.threshold_pub_key,
+        ),
+    };
+    let mut builder = TransactionBuilder::default();
+    builder.output(Output::LN(ContractOrOfferOutput::Offer(offer.clone())));
+    let tx = builder.build(&secp(), &mut rng());
+    fed.submit_transaction(tx);
+
+    // Run 2 epochs to get the offer to show up p
+    fed.run_consensus_epochs(2).await;
+
+    // Gateway escrows ecash to trigger preimage decryption by the federation
+    let (_, contract_id) = gateway
+        .server
+        .buy_preimage_offer(&payment_hash, &amount, rng())
+        .await
+        .unwrap();
+
+    // Now everyone has 0 ecash
+    assert_eq!(user.total_coins(), sats(0));
+    assert_eq!(gateway.user.total_coins(), sats(0));
+
+    fed.run_consensus_epochs(5).await;
+
+    // User gets error when they try to claim gateway's escrowed ecash
+    let response = user
+        .client
+        .claim_incoming_contract(contract_id, real_keypair, rng())
+        .await;
+    assert!(response.is_err());
+    let response = user
+        .client
+        .claim_incoming_contract(contract_id, fake_keypair, rng())
+        .await;
+    assert!(response.is_err());
+
+    // Gateway re-claims their funds
+    let _outpoint = gateway
```

Comment:
When does the gateway know they need to re-claim their funds?

Reply:
The user client, on the other hand, will need to poll or something. We'll need to put some effort into that. Similar situation with the peg-ins where you need to poll the block height ...

Reply:
Why isn't that a problem for LN proper? As long as the GW gets the preimage before the HTLC timeout, even if they get it after others on the path, shouldn't they be able to claim their inbound HTLC and be good?